File: TWBAndroid/app/src/main/java/twb/conwaybrian/com/twbandroid/shuoApi/ShuoApiService.java
package twb.conwaybrian.com.twbandroid.shuoApi;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.threeten.bp.LocalDateTime;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
import twb.conwaybrian.com.twbandroid.model.Article;
import twb.conwaybrian.com.twbandroid.model.Comment;
import twb.conwaybrian.com.twbandroid.model.Like;
import twb.conwaybrian.com.twbandroid.model.User;

public class ShuoApiService {
    private ShuoApi shuoApi;
    private Retrofit retrofitArticleExcerptApi;

    private ShuoApiService() {
        URLRetrofitBuilder urlRetrofitBuilder = new URLRetrofitBuilder();
        retrofitArticleExcerptApi = urlRetrofitBuilder.buildretrofit("http://34.80.143.220:9999/", true);
        shuoApi = retrofitArticleExcerptApi.create(ShuoApi.class);
    }

    // 獲取實例
    public static ShuoApiService getInstance() {
        return ShuoApiService.SingletonHolder.INSTANCE;
    }

    public void register(@NonNull Observer observer,
                         @NonNull User user, boolean isObserveOnIO) {

        Gson gson = new Gson();
        String json = gson.toJson(user);
        shuoApi
                .register(json)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void viewed(@NonNull Observer observer,
                       @NonNull Article article, boolean isObserveOnIO) {

        String articleId = article.getArticleId();
        shuoApi
                .viewed(articleId)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void login(@NonNull Observer observer, @NonNull User user, boolean isObserveOnIO) {
        String authKey = user.authKey();
        shuoApi
                .login(authKey)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);

    }

    public void deleteUser(@NonNull Observer observer, @NonNull User user, boolean isObserveOnIO) {
        String authKey = user.authKey();
        shuoApi
                .deleteUser(authKey)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);

    }

    public void forgotPassword(@NonNull Observer observer, @NonNull String email, boolean isObserveOnIO) {
        shuoApi.forgotPassword(email)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void getArticlesPrivate(@NonNull Observer observer, @NonNull User user, @NonNull LocalDateTime endDateTime, @NonNull LocalDateTime startDateTime, @NonNull String orderBy, @NonNull int offset, @NonNull int limit, boolean isObserveOnIO) {
        shuoApi.getArticlesPrivate(user.authKey(), endDateTime, startDateTime, orderBy, offset, limit)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void getArticlesPublic(@NonNull Observer observer, @NonNull LocalDateTime endDateTime, @NonNull LocalDateTime startDateTime, @NonNull String orderBy, @NonNull int offset, @NonNull int limit, boolean isObserveOnIO) {
        shuoApi.getArticlesPublic(endDateTime, startDateTime, orderBy, offset, limit)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void searchArticle(@NonNull Observer observer, @NonNull String query, @NonNull int limit, @NonNull int offset, boolean isObserveOnIO) {
        shuoApi.searchArticle(query, limit, offset)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void getUserPostHistory(@NonNull Observer observer, @NonNull User user, boolean isObserveOnIO) {
        String userId = user.getUserId();
        shuoApi.getUserPostHistory(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void postArticle(@NonNull Observer observer, @NonNull User user, @NonNull Article article, boolean isObserveOnIO) {
        Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss").create();
        String json = gson.toJson(article);
        System.out.println(json);
        shuoApi.postArticle(user.authKey(), json)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void like(@NonNull Observer observer, @NonNull User user, @NonNull Like like, boolean isObserveOnIO) {
        Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss").create();
        String json = gson.toJson(like);
        System.out.println(json);
        shuoApi.like(user.authKey(), json)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);

    }

    public void comment(@NonNull Observer observer, @NonNull User user, @NonNull Comment comment, boolean isObserveOnIO) {
        Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss").create();
        String json = gson.toJson(comment);
        System.out.println(json);
        shuoApi.comment(user.authKey(), json)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);

    }

    @Deprecated
    public void getComments(@NonNull Observer observer, @NonNull Article article, boolean isObserveOnIO) {
        String articleId = article.getArticleId();
        shuoApi.getComments(articleId)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void getArticleDataPublic(@NonNull Observer observer, @NonNull Article article, boolean isObserveOnIO) {
        String articleId = article.getArticleId();
        shuoApi.getArticleDataPublic(articleId)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    public void getArticleDataPrivate(@NonNull Observer observer, @NonNull User user, @NonNull Article article, boolean isObserveOnIO) {
        String articleId = article.getArticleId();
        shuoApi.getArticleDataPrivate(user.authKey(), articleId)
                .subscribeOn(Schedulers.io())
                .observeOn(isObserveOnIO ? Schedulers.io() : AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);
    }

    // 創建實例
    private static class SingletonHolder {
        private static final ShuoApiService INSTANCE = new ShuoApiService();
    }
}

File: TWB-ios/TWB-ios/shuoApi/ShuoApiService.swift
//
//  ShuoApiService.swift
//  TWB-ios
//
//  Created by Brian on 2019/5/11.
//  Copyright © 2019 Brian. All rights reserved.
//

import UIKit
import RxAlamofire
import RxSwift
import Alamofire
import SwiftyJSON

class ShuoApiService {
    
    static let instance = ShuoApiService()
    
    private var baseURL="http://34.80.143.220:9999/shuo"
    let disposeBag = DisposeBag()
    private  init() {
    }

    
    
    func login(user:User)-> Observable<(HTTPURLResponse,Data)>{
        let  path = "/checkUserExist"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url, parameters: nil, headers: ["Authorization":user.authKey()])
    }
    func register(user:User) -> Observable<(HTTPURLResponse,Data)>  {
         let  path = "/public/register"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.post, url, parameters:user.toJson()
            ,encoding: JSONEncoding.default, headers: ["Content-Type":"application/json"])
    }
    
    func deleteUser(user:User) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/deleteUser/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.delete, url, parameters: nil, headers: ["Authorization":user.authKey()])
    }

    func forgotPassword(email:String) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/forgotPassword/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url, parameters: ["email":email])
    }
    
  
    func postArticle(user:User,article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/postArticle/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.post, url, parameters:article.toJson()
            ,encoding: JSONEncoding.default, headers: ["Content-Type":"application/json","Authorization":user.authKey()])
    }
    
//    @Headers("Content-Type:application/json")
//    @POST("/shuo/postArticle/")
//    Observable<Response<ResponseBody>> postArticle(@Header("Authorization") String authKey, @Body String s);
    
    
    //some wired
    func viewed(article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/viewed"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.post, url,encoding: article.articleId, headers: ["Content-Type":"application/json"])
    }
//
//    @POST("/shuo/public/viewed")
//    Observable<Response<ResponseBody>> viewed(@Body String articleId);
//
//
    
    
    func like(user:User,like:Like) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/like/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.post, url,parameters: like.toJson(),encoding: JSONEncoding.default, headers: ["Content-Type":"application/json","Authorization":user.authKey()])
    }
    
//    @Headers("Content-Type:application/json")
//    @POST("/shuo/like/")
//    Observable<Response<ResponseBody>> like(@Header("Authorization") String authKey, @Body String s);
//
    
    func comment(user:User,comment:Comment) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/comment/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.post, url,parameters: comment.toJson(),encoding: JSONEncoding.default, headers: ["Content-Type":"application/json","Authorization":user.authKey()])
    }
    
    
//    @Headers("Content-Type:application/json")
//    @POST("/shuo/comment/")
//    Observable<Response<ResponseBody>> comment(@Header("Authorization") String authKey, @Body String s);
    
    
    func getComments(article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/getComments/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["articleId":article.articleId],encoding: URLEncoding.default)
    }
    
//
//    @GET("/shuo/public/getComments/")
//    @Deprecated
//    Observable<Response<JsonArray>> getComments(@Query("articleId") String articleId);
//
    
    func getArticleDataPublic(article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/getArticleData/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["articleId":article.articleId],encoding: URLEncoding.default)
    }
    
//
//    @GET("/shuo/public/getArticleData/")
//    Observable<Response<JsonObject>> getArticleDataPublic(@Query("articleId") String articleId);
//
    func getArticleDataPrivate(user:User,article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/getArticleData/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["articleId":article.articleId],encoding: URLEncoding.default,headers: ["Authorization":user.authKey()])
    }
    
    
//    @GET("/shuo/getArticleData/")
//    Observable<Response<JsonObject>> getArticleDataPrivate(@Header("Authorization") String authKey, @Query("articleId") String articleId);
//
//
    func searchArticle(keyword:String,limit:Int,offset:Int) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/searchArticle/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["keyWord":keyword,"limit":limit,"offset":offset],encoding: URLEncoding.default)
    }
    
//    @GET("/shuo/public/searchArticle/")
//    Observable<Response<JsonArray>> searchArticle(@Query("keyWord") String keyWord, @Query("limit") int limit, @Query("offset") int offset);

    
    func getArticles(user:User,orderBy:String,startTime:Date,endTime:Date,limit:Int,offset:Int) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/getArticles/"
        //
        
        let dateFormatterGet = DateFormatter()
        dateFormatterGet.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"

        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["startTime":dateFormatterGet.string(from: startTime),"endTime":dateFormatterGet.string(from: endTime),"orderBy":orderBy,"limit":limit,"offset":offset],encoding: URLEncoding.default,headers: ["Authorization":user.authKey()])
    }
    

//    @GET("/shuo/public/getArticles/")
//    Observable<Response<JsonArray>> getArticles(@Header("Authorization") String authKey, @Query("endTime") LocalDateTime endDateTime, @Query("startTime") LocalDateTime startGateTime, @Query("orderBy") String orderBy, @Query("offset") int offset, @Query("limit") int limit);
//
    func getUserPostHistory(user:User,article:Article) -> Observable<(HTTPURLResponse,Data)>  {
        let  path = "/public/getUserPostHistory/"
        let url = URL(string:baseURL.appending(path))!
        return requestData(.get, url,parameters: ["authorId":article.articleId],encoding: URLEncoding.default,headers: ["Authorization":user.authKey()])
    }
//
//    @GET("/shuo/public/getUserPostHistory/")
//    Observable<Response<JsonArray>> getUserPostHistory(@Header("Authorization") String authKey, @Query("authorId") String authorId);

    
    
    func test(observer:AnyObserver<Data>)  {
        let urlString = "https://www.douban.com/j/app/radio/channels"
        let url = URL(string:urlString)!
        request(.get, url)
            .data()
            .subscribe(observer).disposed(by: disposeBag)
        
    }
}







